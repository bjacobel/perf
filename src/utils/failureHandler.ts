import chalk from 'chalk';
import dedent from 'dedent';
import { Argv } from 'yargs';

export default (msg: string, err: Error, yargs: Argv): void => {
  console.log(dedent`
    ${chalk.bgRed('Error:')} ${
      msg || (err && err.message ? err.message : 'unknown')
    }

  Correct usage for this command is described below:
  ${chalk`{grey -----------------------------------------------------------------}`}

  ${yargs.help().toString()}\n
  `);
  process.exit(1);
};
