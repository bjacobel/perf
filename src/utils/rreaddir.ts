import path from 'path';
import fsSync from 'fs';
const fs = fsSync.promises;
import { compress } from 'brotli';

export interface DirectoryReport {
  children: DirectoryReport[];
  name: string;
  size: number;
  compressedSize?: number;
}

export const rreaddir = async function(
  dirpath: string,
  top?: string,
  checkCompressedSize = false,
): Promise<DirectoryReport[]> {
  const items = await fs.readdir(dirpath);
  return Promise.all(
    items.map(async (shortName) => {
      const fullName = path.join(dirpath, shortName);
      const relativeName = path.relative(top || dirpath, fullName);

      const stats = await fs.lstat(fullName);

      if (stats.isDirectory()) {
        const children = await rreaddir(
          fullName,
          top || dirpath,
          checkCompressedSize,
        );
        return { name: relativeName, children, size: 0, compressedSize: 0 };
      } else {
        let compressedSize = 0;
        if (checkCompressedSize) {
          const buf = await fs.readFile(fullName);
          const compressedBuf = compress(buf);
          compressedSize = compressedBuf.byteLength;
        }

        return {
          name: relativeName,
          size: stats.size,
          ...(checkCompressedSize ? { compressedSize } : {}),
          children: [],
        };
      }
    }),
  );
};
