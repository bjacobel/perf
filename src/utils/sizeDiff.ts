import Table, { HorizontalAlignment, HorizontalTableRow } from 'cli-table3';
import chalk from 'chalk';

export interface TableEntry {
  name: string;
  size: string | number;
  compressedSize?: string | number;
  diff?: string | number;
  compressedSizeDiff?: string | number;
}

export const prettyKb = (bytes: number): string => {
  if (bytes < 1024) return `${bytes} B`;
  const kb = (1.0 * bytes) / 1024;
  return `${kb.toFixed(2)} kB`;
};

export const prettyDiff = (bytes: number, bytesPrime: number): string => {
  const diff = bytesPrime - bytes;
  if (diff === 0) return chalk.gray('unchanged');
  const grow = diff >= 0;
  return (grow ? chalk.green : chalk.red).call(
    undefined,
    `${grow ? '-' : '+'}${prettyKb(Math.abs(diff))}`,
  );
};

const render = ({
  name,
  size,
  compressedSize,
  diff,
  compressedSizeDiff,
}: TableEntry): HorizontalTableRow => {
  return [
    { content: name },
    { content: size, hAlign: 'right' as HorizontalAlignment },
    ...(compressedSize
      ? [{ content: compressedSize, hAlign: 'right' as HorizontalAlignment }]
      : []),
    ...(diff
      ? [{ content: diff, hAlign: 'right' as HorizontalAlignment }]
      : []),
    ...(compressedSize && diff
      ? [
          {
            content: compressedSizeDiff,
            hAlign: 'right' as HorizontalAlignment,
          },
        ]
      : []),
  ];
};

export const renderTable = (
  baseline: boolean,
  rows: TableEntry[],
  compressedSize = false,
): void => {
  // eslint-disable-next-line
  // @ts-ignore 2345
  const table = new Table({
    head: render(
      Object.assign(
        {
          name: chalk.magenta('name'),
          size: chalk.magenta('size'),
        },
        compressedSize ? { compressedSize: chalk.magenta('brotli size') } : {},
        baseline ? { diff: chalk.magenta('baseline diff') } : {},
        baseline && compressedSize
          ? { compressedSizeDiff: chalk.magenta('brotli baseline diff') }
          : {},
      ),
    ),
  });

  table.push(...rows.map(render));

  console.log(table.toString());
};
