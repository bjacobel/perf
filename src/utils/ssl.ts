import pem from 'pem';
import path from 'path';
import { promises } from 'fs';
const fs = promises;

interface Certificate {
  serviceKey: string;
  clientKey: string;
  csr: string;
  certificate: string;
}

let memoizedCert: Certificate;

const genCert = async (): Promise<Certificate> => {
  if (memoizedCert) return Promise.resolve(memoizedCert);

  return new Promise(async (resolve, reject) => {
    pem.config({
      pathOpenSSL: '/opt/homebrew/opt/openssl@3/bin/openssl',
    });
    pem.createCertificate(
      {
        days: 1,
        selfSigned: false,
        serviceCertificate: await caCrt(),
        serviceKey: await caKey(),
      },
      (err, certificate: Certificate) => {
        if (err) reject(err);
        memoizedCert = certificate;
        resolve(certificate);
      },
    );
  });
};

export const caKey = async (): Promise<string> =>
  fs.readFile(path.join(__dirname, '../../data/caKey.pem'), 'utf-8');
export const caCrt = async (): Promise<string> =>
  fs.readFile(path.join(__dirname, '../../data/caCrt.pem'), 'utf-8');

export const key = async (): Promise<string> => (await genCert()).clientKey;
export const cert = async (): Promise<string> => (await genCert()).certificate;
