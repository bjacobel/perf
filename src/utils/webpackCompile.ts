import {
  Stats,
  Compiler,
  Configuration,
  ConfigurationFactory,
  MultiConfigurationFactory,
} from 'webpack';
import path from 'path';
import merge from 'webpack-merge';

// polymorphism is evil
type AmbiguousConfig =
  | Configuration
  | Configuration[]
  | ConfigurationFactory
  | MultiConfigurationFactory;

// webpack.config.js can have a terrifying array of export types. This normalizes them all to an array.
const getConfig = async (ambi: AmbiguousConfig): Promise<Configuration[]> => {
  const appConfigs: Configuration[] = [];

  if (typeof ambi === 'object') {
    if (ambi.hasOwnProperty('length')) {
      return Promise.resolve(ambi as Configuration[]);
    } else {
      appConfigs.push(ambi as Configuration);
    }
  } else if (typeof ambi === 'function') {
    const factoryResult = await ambi(
      { production: true },
      { mode: 'production' },
    );
    appConfigs.push(...(await getConfig(factoryResult)));
  }

  return appConfigs;
};

export default async (...extraConfig: Configuration[]) => {
  const cwd = process.cwd();

  const appConfigs = await getConfig(
    require(path.join(cwd, 'webpack.config.js')),
  );

  const finalConfigs = appConfigs.map((config) =>
    merge(config, ...extraConfig, {
      optimization: {
        noEmitOnErrors: true,
      },
    }),
  );

  const webpack: (c: Configuration[]) => Compiler = require(path.join(
    cwd,
    'node_modules/webpack',
  ));

  const compiler = webpack(finalConfigs);
  return new Promise((resolve, reject) => {
    compiler.run((err: Error, stats: Stats) => {
      if (err || stats.hasErrors()) {
        reject(err || new Error(stats.toJson().errors.join('\n')));
      } else {
        resolve(stats.toJson().assets);
      }
    });
  });
};
