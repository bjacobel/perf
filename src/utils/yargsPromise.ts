import yargs, { Argv, CommandModule } from 'yargs';

export default async (cmd: CommandModule, args: string = '') => {
  return new Promise((resolve, reject) =>
    yargs
      .command(cmd)
      .fail(reject)
      .parse(args, (err: Error, _argv: Argv, output: string) => {
        if (err) {
          reject(err);
        } else {
          resolve(output);
        }
      }),
  );
};
