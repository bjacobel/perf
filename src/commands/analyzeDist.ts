import { Arguments, CommandBuilder } from 'yargs';
import os from 'os';
import path from 'path';
import fsSync from 'fs';
const fs = fsSync.promises;

import {
  prettyKb,
  prettyDiff,
  renderTable,
  TableEntry,
} from '../utils/sizeDiff';
import { rreaddir, DirectoryReport } from '../utils/rreaddir';

export const command = 'analyze-dist <dir> [compareBaseline] [emitBaseline]';

export const describe =
  'display sizes of built assets in dist, optionally compared to a previous run';

export const builder: CommandBuilder = {
  dir: {
    type: 'string',
  },
  compareBaseline: {
    type: 'boolean',
  },
  emitBaseline: {
    type: 'boolean',
  },
};

const tableEntry = (
  stats: DirectoryReport,
  baseline: DirectoryReport | undefined,
) => {
  const matchingBaseline = (name: string): { size: number } | false => {
    const bail = false;
    if (!baseline) return bail;
    if (!baseline.children) return bail;
    return baseline.children.find((file) => file.name === name) ?? bail;
  };
  return (stats.children ?? [])
    .filter((file) => !/\.map$/.test(file.name))
    .filter((file) => file.size)
    .map(({ name, size }) =>
      Object.assign(
        {
          name,
          size: prettyKb(size),
        },
        matchingBaseline(name)
          ? {
              diff: prettyDiff(
                size,
                (matchingBaseline(name) as { size: number }).size,
              ),
            }
          : {},
      ),
    );
};

const extractSizes = (
  stats: DirectoryReport,
  baseline: DirectoryReport | undefined,
): TableEntry[] => {
  const childSizes: TableEntry[] = (stats.children ?? [])
    .map((child, i) =>
      extractSizes(
        child,
        baseline && baseline.children && baseline.children[i],
      ),
    )
    .reduce((a, b) => [...a, ...b], []);

  return [...tableEntry(stats, baseline), ...childSizes];
};

export const handler = async (args: Arguments): Promise<void> => {
  let baseline: DirectoryReport | undefined;
  const baselinePath = path.join(os.tmpdir(), 'baseline.json');

  if (args.compareBaseline) {
    baseline = JSON.parse(await fs.readFile(baselinePath, 'utf-8'));
  }

  const stats = {
    name: args.dir as string,
    children: await rreaddir(args.dir as string),
    size: 0,
  };

  renderTable(!!baseline, extractSizes(stats, baseline));

  if (args.emitBaseline) {
    await fs.writeFile(baselinePath, JSON.stringify(stats));
    console.log(
      `Emitted baseline stats file. Run next time with --compare-baseline to diff against this point in time.`,
    );
  }
};
