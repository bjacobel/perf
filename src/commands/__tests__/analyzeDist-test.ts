import { mocked } from 'ts-jest/utils';
import { promises } from 'fs';

import * as analyzeDist from '../analyzeDist';
import yargsPromise from '../../utils/yargsPromise';
import { rreaddir, DirectoryReport } from '../../utils/rreaddir';
import directoryReport from './directoryReport.json';

jest.mock('../../utils/rreaddir');
jest.mock('fs', () => ({
  ...jest.requireActual('fs'),
  promises: {
    writeFile: jest.fn(() => Promise.resolve()),
    readFile: jest.fn(() => Promise.resolve('{}')),
  },
}));

describe('analyze-dist command', () => {
  beforeEach(() => {
    console.log = jest.fn();
    mocked(rreaddir).mockReturnValueOnce(
      Promise.resolve(directoryReport as DirectoryReport[]),
    );
  });

  it('reports a table of stats', async () => {
    await yargsPromise(analyzeDist, 'analyze-dist path');

    expect(console.log).toHaveBeenCalledTimes(1);
    expect(mocked(console.log).mock.calls[0][0]).toMatchSnapshot();
  });

  it('writes a baseline if sent --emit-baseline', async () => {
    await yargsPromise(analyzeDist, 'analyze-dist path --emit-baseline');

    expect(promises.writeFile).lastCalledWith(
      expect.any(String),
      JSON.stringify({
        name: 'path',
        children: directoryReport,
        size: 0,
      }),
    );
  });

  it('reports brotli sizes if sent --brotli', async () => {
    await yargsPromise(analyzeDist, 'analyze-dist path --brotli');

    expect(console.log).toHaveBeenCalledTimes(1);
    expect(mocked(console.log).mock.calls[0][0]).toMatchSnapshot();
  });

  it('reports a table of stats compared to a baseline', async () => {
    mocked(promises.readFile).mockReturnValueOnce(
      Promise.resolve(
        JSON.stringify({
          name: 'path',
          children: directoryReport,
          size: 0,
        }),
      ),
    );

    await yargsPromise(analyzeDist, 'analyze-dist path --compare-baseline');

    expect(mocked(console.log).mock.calls[0][0]).toMatchSnapshot();
  });
});
