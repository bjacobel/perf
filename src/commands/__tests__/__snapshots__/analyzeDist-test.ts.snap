// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`analyze-dist command reports a table of stats (including compressed) compared to a baseline 1`] = `
"┌──────────────────────────────────────────┬────────────┬─────────────┬───────────────┬──────────────────────┐
│ name                                     │       size │ brotli size │ baseline diff │ brotli baseline diff │
├──────────────────────────────────────────┼────────────┼─────────────┼───────────────┼──────────────────────┤
│ bpm.link.json                            │   96.10 kB │     5.70 kB │     unchanged │            unchanged │
├──────────────────────────────────────────┼────────────┼─────────────┼───────────────┼──────────────────────┤
│ prebuilt_recursive_static_conf.json      │    1.27 kB │       437 B │     unchanged │            unchanged │
├──────────────────────────────────────────┼────────────┼─────────────┼───────────────┼──────────────────────┤
│ raw_static_conf.json                     │      825 B │       312 B │     unchanged │            unchanged │
├──────────────────────────────────────────┼────────────┼─────────────┼───────────────┼──────────────────────┤
│ static_conf.json                         │      876 B │       337 B │     unchanged │            unchanged │
├──────────────────────────────────────────┼────────────┼─────────────┼───────────────┼──────────────────────┤
│ dist/assets.manifest.json                │    1.84 kB │       432 B │     unchanged │            unchanged │
├──────────────────────────────────────────┼────────────┼─────────────┼───────────────┼──────────────────────┤
│ dist/prebuilt_recursive_static_conf.json │    1.27 kB │       437 B │     unchanged │            unchanged │
├──────────────────────────────────────────┼────────────┼─────────────┼───────────────┼──────────────────────┤
│ dist/rhumb-config.json                   │       97 B │        72 B │     unchanged │            unchanged │
├──────────────────────────────────────────┼────────────┼─────────────┼───────────────┼──────────────────────┤
│ dist/static_conf.json                    │      876 B │       337 B │     unchanged │            unchanged │
├──────────────────────────────────────────┼────────────┼─────────────┼───────────────┼──────────────────────┤
│ dist/bundles/project.js                  │ 1189.92 kB │   170.82 kB │     unchanged │            unchanged │
├──────────────────────────────────────────┼────────────┼─────────────┼───────────────┼──────────────────────┤
│ dist/html/index.html                     │   41.00 kB │    12.10 kB │     unchanged │            unchanged │
├──────────────────────────────────────────┼────────────┼─────────────┼───────────────┼──────────────────────┤
│ dist/html/preload-dlb.html               │      305 B │       151 B │     unchanged │            unchanged │
├──────────────────────────────────────────┼────────────┼─────────────┼───────────────┼──────────────────────┤
│ dist/lang/en.js                          │    6.77 kB │     1.98 kB │     unchanged │            unchanged │
├──────────────────────────────────────────┼────────────┼─────────────┼───────────────┼──────────────────────┤
│ dist/sass/project.css                    │  228.33 kB │    24.42 kB │     unchanged │            unchanged │
├──────────────────────────────────────────┼────────────┼─────────────┼───────────────┼──────────────────────┤
│ dist/test/specs.js                       │ 3397.97 kB │   532.21 kB │     unchanged │            unchanged │
├──────────────────────────────────────────┼────────────┼─────────────┼───────────────┼──────────────────────┤
│ dist/test/test.html                      │    2.28 kB │       612 B │     unchanged │            unchanged │
└──────────────────────────────────────────┴────────────┴─────────────┴───────────────┴──────────────────────┘"
`;

exports[`analyze-dist command reports a table of stats 1`] = `
"┌──────────────────────────────────────────┬────────────┐
│ name                                     │       size │
├──────────────────────────────────────────┼────────────┤
│ bpm.link.json                            │   96.10 kB │
├──────────────────────────────────────────┼────────────┤
│ prebuilt_recursive_static_conf.json      │    1.27 kB │
├──────────────────────────────────────────┼────────────┤
│ raw_static_conf.json                     │      825 B │
├──────────────────────────────────────────┼────────────┤
│ static_conf.json                         │      876 B │
├──────────────────────────────────────────┼────────────┤
│ dist/assets.manifest.json                │    1.84 kB │
├──────────────────────────────────────────┼────────────┤
│ dist/prebuilt_recursive_static_conf.json │    1.27 kB │
├──────────────────────────────────────────┼────────────┤
│ dist/rhumb-config.json                   │       97 B │
├──────────────────────────────────────────┼────────────┤
│ dist/static_conf.json                    │      876 B │
├──────────────────────────────────────────┼────────────┤
│ dist/bundles/project.js                  │ 1189.92 kB │
├──────────────────────────────────────────┼────────────┤
│ dist/html/index.html                     │   41.00 kB │
├──────────────────────────────────────────┼────────────┤
│ dist/html/preload-dlb.html               │      305 B │
├──────────────────────────────────────────┼────────────┤
│ dist/lang/en.js                          │    6.77 kB │
├──────────────────────────────────────────┼────────────┤
│ dist/sass/project.css                    │  228.33 kB │
├──────────────────────────────────────────┼────────────┤
│ dist/test/specs.js                       │ 3397.97 kB │
├──────────────────────────────────────────┼────────────┤
│ dist/test/test.html                      │    2.28 kB │
└──────────────────────────────────────────┴────────────┘"
`;

exports[`analyze-dist command reports a table of stats compared to a baseline 1`] = `
"┌──────────────────────────────────────────┬────────────┬───────────────┐
│ name                                     │       size │ baseline diff │
├──────────────────────────────────────────┼────────────┼───────────────┤
│ bpm.link.json                            │   96.10 kB │     unchanged │
├──────────────────────────────────────────┼────────────┼───────────────┤
│ prebuilt_recursive_static_conf.json      │    1.27 kB │     unchanged │
├──────────────────────────────────────────┼────────────┼───────────────┤
│ raw_static_conf.json                     │      825 B │     unchanged │
├──────────────────────────────────────────┼────────────┼───────────────┤
│ static_conf.json                         │      876 B │     unchanged │
├──────────────────────────────────────────┼────────────┼───────────────┤
│ dist/assets.manifest.json                │    1.84 kB │     unchanged │
├──────────────────────────────────────────┼────────────┼───────────────┤
│ dist/prebuilt_recursive_static_conf.json │    1.27 kB │     unchanged │
├──────────────────────────────────────────┼────────────┼───────────────┤
│ dist/rhumb-config.json                   │       97 B │     unchanged │
├──────────────────────────────────────────┼────────────┼───────────────┤
│ dist/static_conf.json                    │      876 B │     unchanged │
├──────────────────────────────────────────┼────────────┼───────────────┤
│ dist/bundles/project.js                  │ 1189.92 kB │     unchanged │
├──────────────────────────────────────────┼────────────┼───────────────┤
│ dist/html/index.html                     │   41.00 kB │     unchanged │
├──────────────────────────────────────────┼────────────┼───────────────┤
│ dist/html/preload-dlb.html               │      305 B │     unchanged │
├──────────────────────────────────────────┼────────────┼───────────────┤
│ dist/lang/en.js                          │    6.77 kB │     unchanged │
├──────────────────────────────────────────┼────────────┼───────────────┤
│ dist/sass/project.css                    │  228.33 kB │     unchanged │
├──────────────────────────────────────────┼────────────┼───────────────┤
│ dist/test/specs.js                       │ 3397.97 kB │     unchanged │
├──────────────────────────────────────────┼────────────┼───────────────┤
│ dist/test/test.html                      │    2.28 kB │     unchanged │
└──────────────────────────────────────────┴────────────┴───────────────┘"
`;

exports[`analyze-dist command reports brotli sizes if sent --brotli 1`] = `
"┌──────────────────────────────────────────┬────────────┬─────────────┐
│ name                                     │       size │ brotli size │
├──────────────────────────────────────────┼────────────┼─────────────┤
│ bpm.link.json                            │   96.10 kB │     5.70 kB │
├──────────────────────────────────────────┼────────────┼─────────────┤
│ prebuilt_recursive_static_conf.json      │    1.27 kB │       437 B │
├──────────────────────────────────────────┼────────────┼─────────────┤
│ raw_static_conf.json                     │      825 B │       312 B │
├──────────────────────────────────────────┼────────────┼─────────────┤
│ static_conf.json                         │      876 B │       337 B │
├──────────────────────────────────────────┼────────────┼─────────────┤
│ dist/assets.manifest.json                │    1.84 kB │       432 B │
├──────────────────────────────────────────┼────────────┼─────────────┤
│ dist/prebuilt_recursive_static_conf.json │    1.27 kB │       437 B │
├──────────────────────────────────────────┼────────────┼─────────────┤
│ dist/rhumb-config.json                   │       97 B │        72 B │
├──────────────────────────────────────────┼────────────┼─────────────┤
│ dist/static_conf.json                    │      876 B │       337 B │
├──────────────────────────────────────────┼────────────┼─────────────┤
│ dist/bundles/project.js                  │ 1189.92 kB │   170.82 kB │
├──────────────────────────────────────────┼────────────┼─────────────┤
│ dist/html/index.html                     │   41.00 kB │    12.10 kB │
├──────────────────────────────────────────┼────────────┼─────────────┤
│ dist/html/preload-dlb.html               │      305 B │       151 B │
├──────────────────────────────────────────┼────────────┼─────────────┤
│ dist/lang/en.js                          │    6.77 kB │     1.98 kB │
├──────────────────────────────────────────┼────────────┼─────────────┤
│ dist/sass/project.css                    │  228.33 kB │    24.42 kB │
├──────────────────────────────────────────┼────────────┼─────────────┤
│ dist/test/specs.js                       │ 3397.97 kB │   532.21 kB │
├──────────────────────────────────────────┼────────────┼─────────────┤
│ dist/test/test.html                      │    2.28 kB │       612 B │
└──────────────────────────────────────────┴────────────┴─────────────┘"
`;
